src/AbstractDependencyRewriter.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+/**
+ * @see       https://github.com/laminas/laminas-dependency-plugin for the canonical source repository
+ * @copyright https://github.com/laminas/laminas-dependency-plugin/blob/master/COPYRIGHT.md
+ * @license   https://github.com/laminas/laminas-dependency-plugin/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Laminas\DependencyPlugin;
+
+use Composer\Composer;
+use Composer\Installer\PackageEvent;
+use Composer\IO\IOInterface;
+use Composer\Plugin\PreCommandRunEvent;
+
+use function array_map;
+use function array_shift;
+use function count;
+use function get_class;
+use function in_array;
+use function preg_split;
+use function sprintf;
+
+abstract class AbstractDependencyRewriter implements RewriterInterface
+{
+    /** @var Composer */
+    protected $composer;
+
+    /** @var IOInterface */
+    protected $io;
+
+    /**
+     * @var Replacements
+     */
+    private $replacements;
+
+    public function __construct()
+    {
+        $this->replacements = new Replacements();
+    }
+
+    public function activate(Composer $composer, IOInterface $io)
+    {
+        $this->composer = $composer;
+        $this->io = $io;
+        $this->output(sprintf('<info>Activating %s</info>', get_class($this)), IOInterface::DEBUG);
+    }
+
+    /**
+     * When a ZF package is requested, replace with the Laminas variant.
+     *
+     * When a `require` operation is requested, and a ZF package is detected,
+     * this listener will replace the argument with the equivalent Laminas
+     * package. This ensures that the `composer.json` file is written to
+     * reflect the package installed.
+     */
+    public function onPreCommandRun(PreCommandRunEvent $event)
+    {
+        $this->output(
+            sprintf(
+                '<info>In %s::%s</info>',
+                get_class($this),
+                __FUNCTION__
+            ),
+            IOInterface::DEBUG
+        );
+
+        if (! in_array($event->getCommand(), ['require', 'update'], true)) {
+            // Nothing to do here.
+            return;
+        }
+
+        $input = $event->getInput();
+        if (! $input->hasArgument('packages')) {
+            return;
+        }
+
+        $input->setArgument(
+            'packages',
+            array_map([$this, 'updatePackageArgument'], $input->getArgument('packages'))
+        );
+    }
+
+    abstract public function onPrePackageInstallOrUpdate(PackageEvent $event);
+
+    /**
+     * @param string $message
+     * @param int $verbosity
+     */
+    protected function output($message, $verbosity = IOInterface::NORMAL)
+    {
+        $this->io->write($message, $newline = true, $verbosity);
```

Comment:
`$newline` var is no need to be defined ? should can be: 

```suggestion
        $this->io->write($message, true, $verbosity);
```

Reply:
Copy & paste from current code